src/Workspaces/Core/Portable/Workspace/Solution/StateChecksums.cs

```diff
@@ -98,16 +98,6 @@ public static SolutionStateChecksums Deserialize(ObjectReader reader)
             result[FrozenSourceGeneratedDocumentText] = await SerializableSourceText.FromTextDocumentStateAsync(state.FrozenSourceGeneratedDocumentState, cancellationToken).ConfigureAwait(false);
         }
 
-        if (searchingChecksumsLeft.Remove(Projects.Checksum))
-        {
-            result[Projects.Checksum] = Projects;
-        }
```

Comment:
this it he meat of the change.  The prior logic was simply crazy, and deeply complex confusing.  This it the logic that says "hey... you give me a checksum, and i will find the corresponding real object associated with it in the actual solution/project/document state.

*except*, lines like thsi were basically saying "oh... your checksum itself refers to a checksum object (like a SolutionChecksums/ProjectChecksums/DocumentChecksums instance), so i'll add my own checksum child back into the result.

This was one of the source of multiple passes and expensiveness with checksums.  When looking up project info, we'd both recurse into child projects as real 'Project' instances, and then again as checksum instances.  Insanity.

All of htat is gone now.  we just pass in checksums, and we get real workspace objects back out.

Reply:
Is there an incoming PR that will make it so that this doesn't do the depth first search? (as you mentioned in the office the other day)

Reply:
Yes. That comes in a follow-up